synapse/http/proxyagent.py

```diff
@@ -144,20 +144,23 @@ def __init__(
         if federation_proxies:
             endpoints = []
             for federation_proxy in federation_proxies:
-                endpoint = HostnameEndpoint(
-                    self.proxy_reactor,
-                    federation_proxy.host,
-                    federation_proxy.port,
-                )
-
-                if federation_proxy.tls:
-                    tls_connection_creator = self._policy_for_https.creatorForNetloc(
+                if isinstance(federation_proxy, InstanceTcpLocationConfig):
```

Comment:
I went ahead and just did the work, as this has to do with the `instance_map` and is sort of tangent on replication working correctly. Since this is the place where it's decided on startup exactly which workers are designated to handle proxying requests, I feel this is the only place that needs touched. Everything else should be for outbound requests that will only be going over internet/lan and not to a local machine.

I did test this theory by hardwiring Complement([4f079d8](https://github.com/matrix-org/synapse/pull/15708/commits/4f079d85b58ec792fbb4ad58c57453332b416518)) to verify it functions as expected (and [reverted](https://github.com/matrix-org/synapse/pull/15708/commits/9f41f3e56f7b9c600f9da8a26628c7959a9a7d00) that change in advance), the test run where it works is [here](https://github.com/matrix-org/synapse/actions/runs/5481164667). Recall that this only tests worker mode in Complement and therefore the Sytest failures appear to be flakes.